### ISSUE TITLE
Map WSTG to ASVS and determine coverage and possible knowledge gaps

### ISSUE THREAD

**Issue by RiieCco** [OWASP/wstg#261]
**What would you like added?**
See how the testing guide content correlates to ASVS controls and determine effective coverage. This will help ASVS users to get more context with the controls. This can than also be used in SKF when generating requirements

**Would you like to be assigned to this issue?**
Check the box if you will submit a PR to add the proposed content. Please read CONTRIBUTING.md.
- [x] Assign me, please!

**Comment by RiieCco** [comment]
https://docs.google.com/spreadsheets/d/1CPaeT1bCoI7OydbNJaIVb4i9QEgyxMydGqSiBSK0aKk

**Comment by kingthorin** [comment]
This kind of requires ASVS to formalize their reference standard: https://github.com/OWASP/ASVS/issues/715

**Comment by ThunderSon** [comment]
The above doc kind of shifted gears and is actually building the full CRE repository, and inside it there is a direct mapping between ASVS and WSTG and other projects.

**Comment by kingthorin** [comment]
What I was getting at is that if we are going to reference ASVS in WSTG we need a solid way to do it.

**Comment by themayursinha** [comment]
Is it going to be released in WSTG v5?

**Comment by ThunderSon** [comment]
Hello @themayursinha!
This task is somewhat going to be a bit bigger. Since we since such a huge opportunity out of this, a project is currently being run in parallel to map out requirements, test guides, code advice, standards, policies, etc.
This project is the [Integration Standards Project](https://owasp.org/www-project-integration-standards/). We are looking to have some sort of an MVP in the summer. A lot of thought is going into it.

All projects will be affected and linked under a certain umbrella ID, and that ID will create the maps underneath it.

In short, there is work happening to make that happen :)

**Comment by ThunderSon** [comment]
As this is the work of another project, this will be closed and tracking should follow with the other project :)

**Comment by jeremychoi** [comment]
While this was closed and there was the other project, I've created https://github.com/jeremychoi/owasp-asvs-wstg-checklist which would be relevant to this issue.

**Comment by ThunderSon** [comment]
@jeremychoi this is different! I love it. @kingthorin this is something we should look into taking in.

Why this is different?
This allows the attacker and the reviewer to understand the level of coverage, and their stance overall, which is different from simply mapping everything together!
The new project will map things out, but not give smart information (yet) :)

I'll await Rick's comments, once done, if in agreement, create a PR to add the XLSX to this repository!

Thanks :)

**Comment by jeremychoi** [comment]
@ThunderSon  I see. Thanks. If the files could be added to this repo, that would be great.  One thing I am not sure about is if there is something to be done with regards to the license(MIT) of the spreadsheet file. I created them based on https://github.com/shenril/owasp-asvs-checklist and added the WSTG mapping information to it. Your help would be appreciated on that.

**Comment by kingthorin** [comment]
Sure I guess I'm fine with it being added as a checklist artifact. It would be really nice if it was a non-proprietary format like tsv, csv, etc instead of XLS/XLSX though. If it has to be maintained/offered as an Excel file then it should be done similar to the existing one (specifying the hash and other info).

**Comment by jeremychoi** [comment]
Thanks for the comment. I'll create a csv one soon.

--
Jeremy Choi / Red Hat Product Security

**Comment by ThunderSon** [comment]
@jeremychoi this is not critical nor urgent. Actually this can wait enough till v5 is being prepared.
Since you poked at another issue, #492 , that one is definitely more critical if we can focus on it.

**Comment by themayursinha** [comment]
hi, I've been wondering if this is already applied in the new version?

**Comment by kingthorin** [comment]
The issue would be closed if the work was done :wink:

**Comment by ryarmst** [comment]
If this is still intended, I did some work on ASVS V5 and would be happy to work on mapping.

**Comment by prasunsrivastav123-lang** [comment]
@kingthorin is it still in progress?

**Comment by kingthorin** [comment]
I haven't looked but this may have already been tackled via OpenCRE.

**Comment by prasunsrivastav123-lang** [comment]
@kingthorin Thanks for the heads-up. I'll check the existing OpenCRE mapping for WSTG-ASVS coverage and gaps, and follow up if there's still work needed on our side.

**Comment by avp9-nexus** [comment]
Following up on the OpenCRE question from March - I could not find the follow-up in the thread, so I measured it. Everything below is reproducible from two public endpoints: `https://www.opencre.org/rest/v1/standard/<name>` and `https://www.opencre.org/rest/v1/map_analysis?standard=<a>&standard=<b>`.

**OpenCRE does carry a WSTG-ASVS mapping, but it covers roughly half the guide, against a superseded ASVS.**

- OpenCRE's WSTG catalog holds **56 test IDs**, hyperlinked to v4.2 pages. The v4.2 checklist has **97 test IDs**, so **42 of them have no OpenCRE entry at all** - and the absences are not evenly spread: the **entire Information Gathering chapter is missing (10/10)**, business logic loses 6 of 9, input validation 6, client-side and configuration 5 each. Business logic - the chapter new proposals here keep getting routed into (#1262, #1267) - is mostly invisible to OpenCRE.
- OpenCRE's ASVS entries hyperlink to **v4.0.3**, which v5.0.0 superseded in May 2025. Any coverage answer drawn from OpenCRE today is an answer about ASVS 4.0.3.
- Within what is mapped: 54 of the 56 WSTG entries reach at least one ASVS control (the two that reach none: WSTG-CONF-02 and WSTG-ERRH-02). In the other direction, 172 of the 277 ASVS controls loaded in OpenCRE reach no WSTG test - though a large share of those are architecture and documentation controls (starting with the whole V1 chapter) that a black-box testing guide would not be expected to cover, so the raw number overstates the real knowledge gap.

So: partially tackled. A current mapping - guide master against ASVS 5.0 - does not exist in OpenCRE today, and the WSTG side of what does exist stops at 56 entries. Whether that gap is worth closing here or upstream in OpenCRE is your call - happy to post the raw per-ID diff if useful.

**Comment by kingthorin** [comment]
This is now actually documented in the summary of every checklist deployment. Ex: https://github.com/OWASP/wstg/actions/runs/33028843495

The JSON checklist does have a cre-ids array where we were able to map things via the API, ex:
```json

  "cre_ids": [
            "843-841",
            "402-706"
          ]
```

**Comment by avp9-nexus** [comment]
That summary is better than the diff I offered, since it regenerates per deployment and already carries the per-test status - so scratch that offer. For what it is worth, reading `checklists/checklist.json` at master gives 55 of 113 tests carrying a non-empty `cre_ids`, with the unmapped 58 concentrated in Information Gathering (10 of 10), Input Validation (10) and Business Logic (7).

One caveat on my own numbers first: your generator queries the same `/rest/v1/standard` endpoint I did, so our figures agreeing confirms I read OpenCRE correctly, not that the mapping itself is independently corroborated. Same source, read twice.

The part that is still open is the one the issue title asks for. `ASVS` appears nowhere in `generate_checklist_json.py` or `build-checklists.yml` - the pipeline resolves WSTG to CRE ids and stops there. The ASVS side exists only inside OpenCRE, and there it is uniformly 4.0.3: across all 14 pages of `/rest/v1/standard/ASVS`, 64 distinct hyperlinks, every one pinned to `ASVS/blob/v4.0.3`. ASVS 5.0.0 shipped in May 2025.

So the chain today answers "which CRE", and through CRE "which 4.0.3 control" - not which 5.0 control. That gap sits upstream in OpenCRE rather than in this repo, which may well make it out of scope here; but it does mean a coverage answer derived from `cre_ids` is currently an answer about the previous major version of ASVS.

**Comment by kingthorin** [comment]
The reason we may only cre_ids thus far as that I figured it was a better central location to tie things together, instead of trying to map WSTG to _**n**_ other standards/guides (that would end-up duplicating too much effort).

I'm also still unsure if we should include any mapping int he guide itself, because where do we draw the line?
- CWEs?
- CREs?
- CAPEC?
- ASVS?
- Top 10?
- WASC?
- Bug Bounty Taxonomy? (Is there a universal one?)

**Comment by avp9-nexus** [comment]
The hub reasoning holds up when measured, which is worth having on the record rather than as an intuition. Through the CRE hub, the 56 WSTG entries currently in OpenCRE reach ASVS for 54 of them, CWE for 52, and CAPEC for 44. One mapping, three standards, at 79 to 96 percent. Mapping each of those directly would be three times the work for the same reach.

Your list also answers itself on where the line falls. Of the seven you name, OpenCRE carries CWE, CAPEC, ASVS, OWASP Top 10 2021 and CRE itself. WASC and a bug bounty taxonomy are not in it at all - and for the second one your parenthetical is the reason, since there is no universal taxonomy for the hub to have adopted. So the line is not really a judgement call: the hub already draws it, and it moves when the hub moves rather than when the guide does.

One measured oddity from the same pull, since it points at the hub rather than at you: the WSTG to OWASP Top 10 2021 path resolves for only 2 of the 56 entries. That is an OWASP to OWASP link, and it is by far the thinnest of the four - CWE and ASVS both sit above 90 percent. If the guide ever wants a Top 10 cross-reference, `cre_ids` will not supply it today.

None of which argues for putting mappings in the guide text. It argues that the one you already store is doing more work than it looks like it is.
